sw/device/lib/testing/usb_testutils_streams.c

```diff
@@ -0,0 +1,488 @@
+  s->tx_bytes = 0u;
+  s->rx_bytes = 0u;
+  s->transfer_bytes = transfer_bytes;
+
+  // Initialize the LFSR state for transmission and reception sides
+  // - we use a simple LFSR to generate a PRND stream to transmit to the USBPI
+  // - the USBDPI XORs the received data with another LFSR-generated stream of
+  //   its own, and transmits the result back to us
+  // - to check the returned data, our reception code mimics both LFSRs
+  s->tx_lfsr = USBTST_LFSR_SEED(id);
+  s->rxtx_lfsr = s->tx_lfsr;
+  s->rx_lfsr = USBDPI_LFSR_SEED(id);
+
+  // Packet size randomization
+  s->tx_buf_size = BUFSZ_LFSR_SEED(id);
+
+  // Set up the endpoint for IN transfers (TO host)
+  //
+  // Note: We install the rx_show handler to catch any misdirected data
+  // transfers
+  void (*rx)(void *, dif_usbdev_rx_packet_info_t, dif_usbdev_buffer_t) =
+      (ep_in == ep_out) ? strm_rx : rx_show;
+
+  s->tx_ep = ep_in;
+  usb_testutils_endpoint_setup(ctx->usbdev, ep_in, kUsbdevOutStream, s,
+                               strm_tx_done, rx, NULL, NULL);
+  s->rx_ep = ep_out;
+  if (ep_out != ep_in) {
+    // Set up the endpoint for OUT transfers (FROM host)
+    usb_testutils_endpoint_setup(ctx->usbdev, ep_out, kUsbdevOutStream, s, NULL,
+                                 strm_rx, NULL, NULL);
+  }
+}
+
+// Service the given stream, preparing and/or sending any data that we can;
+// data reception is handled via callbacks and requires no attention here
+void usb_testutils_stream_service(usb_testutils_streams_ctx_t *ctx,
+                                  uint8_t id) {
+  // Locate the stream context information
+  CHECK(id < USBUTILS_STREAMS_MAX);
+  usbdev_stream_t *s = &ctx->streams[id];
+
+  // Generate output data as soon as possible and make it available for
+  //   collection by the host
+
+  uint8_t tx_ep = s->tx_ep;
+  uint8_t nqueued = ctx->tx_bufs_queued[tx_ep];
+
+  if (s->tx_bytes < s->transfer_bytes &&      // More bytes to transfer?
+      nqueued < ctx->tx_bufs_limit[tx_ep] &&  // Endpoint allowed buffer?
+      ctx->tx_queued_total <
+          USBUTILS_STREAMS_TXBUF_MAX) {  // Total buffers not exceeded?
+    dif_usbdev_buffer_t buf;
+
+    // See whether we can populate another buffer yet
+    dif_result_t dif_result = dif_usbdev_buffer_request(
+        ctx->usbdev->dev, ctx->usbdev->buffer_pool, &buf);
+    if (dif_result == kDifOk) {
+      // This is just for reporting the number of buffers presented to the
+      // USB device, as a progress indicator
+      static unsigned bufs_sent = 0u;
+      uint32_t num_bytes;
+
+      if (s->sent_sig) {
+        if (s->flags & kUsbdevStreamFlagMaxPackets) {
+          num_bytes = USBDEV_MAX_PACKET_SIZE;
+        } else {
+          // Vary the amount of data sent per buffer
+          num_bytes = s->tx_buf_size % (USBDEV_MAX_PACKET_SIZE + 1u);
+          s->tx_buf_size = LFSR_ADVANCE(s->tx_buf_size);
+        }
+        uint32_t tx_left = s->transfer_bytes - s->tx_bytes;
+        if (num_bytes > tx_left)
```

Comment:
As per the [style guide](https://opentitan.org/book/doc/contributing/style_guides/c_cpp_coding_style.html#formatting-of-loops-and-conditionals), single line conditionals and loops must use braces.

Reply:
I'm somewhat surprised that clang_format_fix doesn't modify that, but thank you for catching it.